<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-del-projecto-team-1/blob/main/Semana3_Actividad1_Equipo1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Maestría en Inteligencia Artificial Aplicada**
##**Curso: Ciencia y analítica de datos**
###Tecnológico de Monterrey
### **Semana 3** - **Actividad 1.**
###Prof Jobish Vallikavungal Devassia
<br>

###Equipo 1
###Alumno: Joel Orlando Hernández Ramos
###Matrícula: A00759664

###Alumno: Rafael Sánchez García
###Matrícula: A01182780

Fecha de Entrega: Octubre 04, 2021


##**Parte 1: Fundamentos de bases de datos**

####1. Fundaments of Data bases for data science

We can define data base as a set of data with an specific structure where this data is about an activity in the real world and it is stored as a data in a computer. This database has a particular purpose like storage the data about a health, credit cards, customer etc.

The reason of why this database should be structured is because it can be access fast and easy with a simply query, it can provide quick insights without too much effort and keep data secure.

When a database is designed should be take in consideration the purpose of it to define which type of database should be chosen, following the CAP (Consistency, Availability, Partition tolerance) theorem, where should be prioritized at least 2 of its characteristics. 

There are 2 types of database:
1. Relational database. Which is a structured database storage in different tables linked between them.
2. Non-relational database. Which is a database where is stored by categories and not by relation, to put more in context lets think like a Python dictionary where we have keys (should be unique).

The main difference between both is that Relational database has rows and columns, it is stored in tables and can be linked between different tables, in the other hand the non Relational database is storage in a structure document with Keys

The database is managed by the DBMS (Data Base Management System), which is a system that can be used to create, delete, control data depending a request or query.

####2. Fundaments of Data Warehouse for data science
The data warehouse is formed by multiple large data bases from different sources and it is considered also as a data management system (DBMS), where queries (request) can be executed to get data or manipulate de data. 

During the time the warehouse is more valuable because it will storage more info to have historical data. Also is so powerful  allowing companies, institutions and organizations to analyze larga data, get insights and storage historical data.

Some of the benefits are:

- Subject-oriented. It can perform analysis segregating the database in one topic.
- Integrated. It can get the data form different data sources.
- Nonvolatile. This one of the most important benefits, because the data keeps stable and never change by itself.
- Time- variant. It can be perform analysis in a historical point of view.

As the data base the warehouse should be designed with a good structure and format, understand the business requirements and scope, it should establish relationship between different groups of data, have a good environment to support the warehouse and define how often the data will be refresh.

It is considered as a central repository of information, where based on the data desicions can be made.



##**Reference:**

Amazon. (s.f) Data Warehouse Concepts. AWS. Recuperado Octubre 3 2022 de:
https://aws.amazon.com/data-warehouse/

Metwalli, Sara (2020). Databases 101: Introduction to Database for Data Scientist. Towards Data Science. Recuperado Octubre 3 2022 de: https://towardsdatascience.com/databases-101-introduction-to-databases-for-data-scientists-ee18c9f0785d


##**Parte 2: Selección y limpieza de los Datos en Python**

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns


1. We will import the dataframe provided by the Professor.

In [ ]:
input='https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv'

dataframe= pd.read_csv(input)
dataframe.head(10)  #this will help us visualize the data. A preview

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...       X15  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...       0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...    3272.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...   14331.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...   28314.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...   20940.0   
5   6   50000  1.0  1.0  2.0  37.0  0.0  0.0  0.0  0.0  ...   19394.0   
6   7  500000  1.0  1.0  2.0  29.0  0.0  0.0  0.0  0.0  ...  542653.0   
7   8  100000  2.0  2.0  2.0  23.0  0.0 -1.0 -1.0  0.0  ...     221.0   
8   9  140000  2.0  3.0  1.0  28.0  0.0  0.0  2.0  0.0  ...   12211.0   
9  10   20000  1.0  3.0  2.0  35.0 -2.0 -2.0 -2.0 -2.0  ...       0.0   

        X16       X17      X18      X19      X20      X21      X22      X23  \
0       0.0       0.0      0.0    689.0      0.0      0.0      0.0      0.0   
1    3455.0    3261.0      0.0   1000.0   1000.0   1000.0      0.0   2000.0   
2   14948.0   15549.0   1518.0   1500.0   1000.0   1000.0   1000.0   5000.0   
3   28959.0   29547.0   2000.0   2019.0   1200.0   1100.0   1069.0   1000.0   
4   19146.0   19131.0   2000.0  36681.0  10000.0   9000.0    689.0    679.0   
5   19619.0   20024.0   2500.0   1815.0    657.0   1000.0   1000.0    800.0   
6  483003.0  473944.0  55000.0  40000.0  38000.0  20239.0  13750.0  13770.0   
7    -159.0     567.0    380.0    601.0      0.0    581.0   1687.0   1542.0   
8   11793.0    3719.0   3329.0      0.0    432.0   1000.0   1000.0   1000.0   
9   13007.0   13912.0      0.0      0.0      0.0  13007.0   1122.0      0.0   

     Y  
0  1.0  
1  1.0  
2  0.0  
3  0.0  
4  0.0  
5  0.0  
6  0.0  
7  0.0  
8  0.0  
9  0.0  

[10 rows x 25 columns]

In [ ]:
dataframe.shape #we run this method to see the size of our array (rows, columns)

(30000, 25)

The first thing that we will do once we imported our dataframe is understand the data that we are working with.

In [ ]:
dataframe.info() #this will provide us a description of the data types per column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

After running the code dataframe info code, we can see that we have 30,000 rows per column. Also we can see that all our data is considered numerical. Which is good for statistical analysis as we won't need to convert data types like strings into numbers. Also, we can see that we have some null values per rows. However, they are really minimal. For example, Column X16 has the most amount of null values but its only 17 out of 30,000. Only .05% of the total sample, which in ou opinion is not really significant if we get rid of those values. However, we will first analyze if we can adjust them in any other way. 

Based on the data set information they gave us. We know that the following columns, all though they are using numerical values, can be considered as categorical. 

* X2 (GENDER)
* X3 (EDUCATION)
* X4 (MARITAL STATUS)
* X6-X11 (PAYMENT STATUS)

For these values we have a range of specific numerical values that should be used and specific categorical values for each numerical value. For example, in gender
1 = male; 2 = female. 

In [ ]:
dataframe[['X2', 'X3', 'X4', 'X6','X7','X8','X9','X10', 'X11']].describe()

X2            X3            X4            X6            X7  \
count  29999.000000  29998.000000  29998.000000  29997.000000  29995.000000   
mean       1.603753      1.853057      1.551903     -0.016635     -0.133689   
std        0.489125      0.790320      0.521968      1.123829      1.197254   
min        1.000000      0.000000      0.000000     -2.000000     -2.000000   
25%        1.000000      1.000000      1.000000     -1.000000     -1.000000   
50%        2.000000      2.000000      2.000000      0.000000      0.000000   
75%        2.000000      2.000000      2.000000      0.000000      0.000000   
max        2.000000      6.000000      3.000000      8.000000      8.000000   

                 X8            X9           X10           X11  
count  29993.000000  29991.000000  29984.000000  29986.000000  
mean      -0.166405     -0.220800     -0.266342     -0.291136  
std        1.196048      1.169153      1.133296      1.150134  
min       -2.000000     -2.000000     -2.000000     -2.000000  
25%       -1.000000     -1.000000     -1.000000     -1.000000  
50%        0.000000      0.000000      0.000000      0.000000  
75%        0.000000      0.000000      0.000000      0.000000  
max        8.000000      8.000000      8.000000      8.000000

The describe function gave us some valuable information for each column. We will start interpretating the information for X2 (GENDER). 

### COLUMN X2 (GENDER)

First, we should notice that the count for this column is detecting 29999 numbers instead of 30,000. We can infer we have one NaN value in this dataset. Also, As we know, for this variable we should have values of only 1 and 2. As we can see, the table is telling us that we have minimum values of 1 and maximum values of 2, which make us infer that the our 29,999 numbers just have those values.

However, we can double check all this info with the following code. 

In [ ]:
dataframe['X2'].unique()

array([ 2.,  1., nan])

With this, we are not are certain that in the Gender Column we have 29,999 values of either 1 or 2 and one NaN. 

Since the mean is 1.6 and we can only use 1.0 or 2.0, we cannot use the mean to substitute the NaN value.  

In [ ]:
dataframe['X2'].median()

2.0

In [ ]:
dataframe['X2'].mode()

0    2.0
dtype: float64

We will substitute the NaN value with 2.0 since both the median and the mode are equal to 2.0.

In [ ]:
dataframe['X2'].replace(np.nan, 2.0, inplace=True )

In [ ]:
dataframe[['X2']].describe()

X2
count  30000.000000
mean       1.603767
std        0.489122
min        1.000000
25%        1.000000
50%        2.000000
75%        2.000000
max        2.000000

Now our code is detecting 30,000 number entries in the Gender Column. 

### COLUMN X3 (EDUCATION)

We will follow a similar process to the one we just did.  However, for this column we should notice that we can only have 4 values 1 = graduate school; 2 = university; 3 = high school; 4 = others. In our table, we can see that we have two NaN,  a max of 6 and a min of 4. Therefore, we have several data that need to be corrected. 

In [ ]:
dataframe['X3'].unique()

array([ 2.,  1.,  3.,  5.,  4.,  6.,  0., nan])

According to our data set information, we have 4 values that should be replaced. 5, 6, 0 and NaN should be adjusted. 

Since, in this case we know that 4 = others, then we believe that all values that are misleading (5, 6, 0 and NaN) should be stored as a 4.

In [ ]:
dataframe['X3'].replace(np.nan, 4.0, inplace=True )
dataframe['X3'].replace(5.0, 4.0, inplace=True )
dataframe['X3'].replace(6.0, 4.0, inplace=True )
dataframe['X3'].replace(0, 4.0, inplace=True )

In [ ]:
dataframe[['X3']].describe()

X3
count  30000.000000
mean       1.842333
std        0.744643
min        1.000000
25%        1.000000
50%        2.000000
75%        2.000000
max        4.000000

In [ ]:
dataframe['X3'].unique()

array([2., 1., 3., 4.])

We now have 30,000 values between 1 and 4. 

### COLUMN X4 (Marital status)

For this column we should only have 3 values 1 = married; 2 = single; 3 = others. Instead, it seems that we have NaN values and 0's 



In [ ]:
dataframe['X4'].unique()

array([ 1.,  2.,  3.,  0., nan])

We will have to replace 0's and NaN. Again, in this case we have a "others" category. Therefore, we will store the misleading data under this category. 

In [ ]:
dataframe['X4'].replace(np.nan, 3.0, inplace=True )
dataframe['X4'].replace(0, 3.0, inplace=True )

In [ ]:
dataframe[['X4']].describe()

X4
count  30000.000000
mean       1.557400
std        0.521518
min        1.000000
25%        1.000000
50%        2.000000
75%        2.000000
max        3.000000

In [ ]:
dataframe['X4'].unique()

array([1., 2., 3.])

We now have 30,000 values between 1 and 3. 

### COLUMNS X6 - X11 (Payment statos from April to September 2005)

All these columns should have the following values -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above.


In [ ]:
print(dataframe['X6'].unique())
print(dataframe['X7'].unique())
print(dataframe['X8'].unique())
print(dataframe['X9'].unique())
print(dataframe['X10'].unique())
print(dataframe['X11'].unique())


[ 2. -1.  0. -2.  1.  3.  4.  8.  7.  5.  6. nan]
[ 2.  0. -1. -2.  3.  5.  7.  4.  1.  6. nan  8.]
[-1.  0.  2. -2.  3.  4.  6.  7.  1.  5. nan  8.]
[-1.  0. -2.  2.  3.  4. nan  5.  7.  6.  1.  8.]
[-2.  0. -1.  2. nan  3.  5.  4.  7.  8.  6.]
[-2.  2.  0. -1.  3.  6.  4.  7.  8. nan  5.]


In [ ]:
print(dataframe['X6'].median())
print(dataframe['X7'].median())
print(dataframe['X8'].median())
print(dataframe['X9'].median())
print(dataframe['X10'].median())
print(dataframe['X11'].median())

0.0
0.0
0.0
0.0
0.0
0.0


In [ ]:
print(dataframe['X6'].mode())
print(dataframe['X7'].mode())
print(dataframe['X8'].mode())
print(dataframe['X9'].mode())
print(dataframe['X10'].mode())
print(dataframe['X11'].mode())

0    0.0
dtype: float64
0    0.0
dtype: float64
0    0.0
dtype: float64
0    0.0
dtype: float64
0    0.0
dtype: float64
0    0.0
dtype: float64


As we can see, in this columns, we have unwanted data such as 0, -2 and NaN. However, we also see that the median and mode of each of this column is always 0, an undefined value. In this regard, we think that it would be convenient to  group all unwanted data (0, -2 and NaN) inside the category 'pay duly', since 0 and -2 probably means that they did not have a delay on the payment. Also, we could define 0 as "pay duly" instead of -1 that way we have a sequence of numbers.

In [ ]:
dataframe['X6'].replace([np.nan,0,-1,-2], 0.0, inplace=True )
dataframe['X7'].replace([np.nan,0,-1,-2], 0.0, inplace=True )
dataframe['X8'].replace([np.nan,0,-1,-2], 0.0, inplace=True )
dataframe['X9'].replace([np.nan,0,-1,-2], 0.0, inplace=True )
dataframe['X10'].replace([np.nan,0,-1,-2], 0.0, inplace=True )
dataframe['X11'].replace([np.nan,0,-1,-2], 0.0, inplace=True )

In [ ]:
dataframe[['X6','X7','X8','X9','X10', 'X11']].describe()

X6            X7            X8            X9           X10  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean       0.356767      0.320033      0.303800      0.258633      0.221367   
std        0.760594      0.801727      0.789341      0.760983      0.717575   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        8.000000      8.000000      8.000000      8.000000      8.000000   

                X11  
count  30000.000000  
mean       0.226500  
std        0.715366  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        8.000000

In [ ]:
print(dataframe['X6'].unique())
print(dataframe['X7'].unique())
print(dataframe['X8'].unique())
print(dataframe['X9'].unique())
print(dataframe['X10'].unique())
print(dataframe['X11'].unique())

[2. 0. 1. 3. 4. 8. 7. 5. 6.]
[2. 0. 3. 5. 7. 4. 1. 6. 8.]
[0. 2. 3. 4. 6. 7. 1. 5. 8.]
[0. 2. 3. 4. 5. 7. 6. 1. 8.]
[0. 2. 3. 5. 4. 7. 8. 6.]
[0. 2. 3. 6. 4. 7. 8. 5.]


Now we only have defined numbers by our data set information. 



---



---



At this point, we have finished cleaning our dataser for the categorical columns. However, based on the data set information they gave us, the remaining columns are not categorical and they have to do mainly with money. 

* X1 (CREDIT AMOUNT)
* X5 (AGE)
* X12-17 (AMOUNT OF BILL STATEMENT)
* X18-X23 (AMOUNT OF PREVIOUS PAYMENTS)

In [ ]:
dataframe[['X1', 'X5', 'X12', 'X13','X14','X15','X16','X17']].describe()

X1            X5            X12            X13  \
count    30000.000000  29995.000000   29989.000000   29989.000000   
mean    167484.322667     35.484214   51236.862750   49190.734669   
std     129747.661567      9.218024   73645.219278   71183.385123   
min      10000.000000     21.000000 -165580.000000  -69777.000000   
25%      50000.000000     28.000000    3565.000000    2986.000000   
50%     140000.000000     34.000000   22387.000000   21207.000000   
75%     240000.000000     41.000000   67139.000000   64027.000000   
max    1000000.000000     79.000000  964511.000000  983931.000000   

                X14            X15            X16            X17  
count  2.998700e+04   29985.000000   29983.000000   29990.000000  
mean   4.702535e+04   43275.652326   40324.493980   38881.135745  
std    6.936086e+04   64345.500073   60809.984983   59561.312967  
min   -1.572640e+05 -170000.000000  -81334.000000 -339603.000000  
25%    2.667500e+03    2329.000000    1763.500000    1256.250000  
50%    2.008900e+04   19052.000000   18107.000000   17081.000000  
75%    6.018200e+04   54560.000000   50213.000000   49208.250000  
max    1.664089e+06  891586.000000  927171.000000  961664.000000

In [ ]:
dataframe[['X18', 'X19', 'X20', 'X21','X22','X23', 'Y']].describe()

X18           X19            X20            X21  \
count   29992.000000  2.999100e+04   29992.000000   29989.000000   
mean     5662.945886  5.922489e+03    5225.623400    4827.252526   
std     16564.165089  2.304418e+04   17608.422625   15668.751975   
min         0.000000  0.000000e+00       0.000000       0.000000   
25%      1000.000000  8.355000e+02     390.000000     296.000000   
50%      2100.000000  2.009000e+03    1800.000000    1500.000000   
75%      5006.000000  5.000000e+03    4505.500000    4014.000000   
max    873552.000000  1.684259e+06  896040.000000  621000.000000   

                 X22            X23             Y  
count   29989.000000   29995.000000  29997.000000  
mean     4800.297209    5216.259977      0.221189  
std     15280.842069   17778.848359      0.415054  
min         0.000000       0.000000      0.000000  
25%       251.000000     118.000000      0.000000  
50%      1500.000000    1500.000000      0.000000  
75%      4033.000000    4000.000000      0.000000  
max    426529.000000  528666.000000      1.000000

### COLUMN X1 (CREDIT)
With the descprition of Column X1, it seems that there is no need to change or clean column A as all data seem to be compplete: there is no NaN values or weird values like a minimum negative value (which would not make sense since it is a credit). 



In [ ]:
dataframe['X1'].isna().any()

False

In [ ]:
dataframe['X1'].isnull().any()

False

With the isnull and isna functions we can double check if we have any null values. They will return a boolean value, where false means there is no null values. 

### COLUMN X5 (AGE)


According to our quick descriptvive analysis, we have five missing values in this column. The rest of the values seem to be correct as the minimum value is 21 (which make sense since it is the legal age in Taiwan) and the max value is 79. 

Since our data set is big enough, we will fill our NaN values with the median as it will not show any bias related problems. 

In [ ]:
dataframe['X5'].median()

34.0

In [ ]:
dataframe['X5'].replace([np.nan], 34.0, inplace=True )

In [ ]:
dataframe['X5'].isnull().any()

False

In [ ]:
dataframe['X5'].describe()

count    30000.000000
mean        35.483967
std          9.217276
min         21.000000
25%         28.000000
50%         34.000000
75%         41.000000
max         79.000000
Name: X5, dtype: float64

Now, we have 30,000 values for our age, the same distribution of our data and almost the same mean and std. 

### COLUMN X12-X17 (AMOUNT OF BILL STATEMENT FROM APRIL TO SEPTEMBER, 2005)


First, it is important to understand that a bill statement is a monthly report sent to a credit card holder showing their information regarding the payment due and other information. According to bibliography and our personal experience, a negative balance is possible due to an overpayment. Therefore, we believe that this information is relevant for the analysis.

As you can see in our descriptive table, we have negative values in the minimum value which means that at least one person did an overpayment. As we mentioned, for this exercise we will keep those numbers. 

In that regard, the only important information that we have to analyze is the missing values. We have NaN values in all of these columns.

We will evaluate if using the median is a good option. We believe it is based on the same logic that Sour data set is big enough, so it will not show any bias related problems. Also we would like to use it to keep consistency in the type of imputation we are using for the COLUMNS with numerical values. 

In [ ]:
print(dataframe['X12'].median())
print(dataframe['X13'].median())
print(dataframe['X14'].median())
print(dataframe['X15'].median())
print(dataframe['X16'].median())
print(dataframe['X17'].median())

22387.0
21207.0
20089.0
19052.0
18107.0
17081.0


In [ ]:
dataframe['X12'].replace([np.nan], dataframe['X12'].median(), inplace=True )
dataframe['X13'].replace([np.nan], dataframe['X13'].median(), inplace=True )
dataframe['X14'].replace([np.nan], dataframe['X14'].median(), inplace=True )
dataframe['X15'].replace([np.nan], dataframe['X15'].median(), inplace=True )
dataframe['X16'].replace([np.nan], dataframe['X16'].median(), inplace=True )
dataframe['X17'].replace([np.nan], dataframe['X17'].median(), inplace=True )

In [ ]:
dataframe[['X12','X13','X14','X15','X16','X17']].describe()

X12            X13           X14            X15  \
count   30000.000000   30000.000000  3.000000e+04   30000.000000   
mean    51226.284467   49180.473967  4.701368e+04   43263.540500   
std     73633.787601   71172.349726  6.934810e+04   64331.690439   
min   -165580.000000  -69777.000000 -1.572640e+05 -170000.000000   
25%      3566.750000    2986.750000  2.674750e+03    2332.000000   
50%     22387.000000   21207.000000  2.008900e+04   19052.000000   
75%     67091.000000   64006.250000  6.016475e+04   54506.000000   
max    964511.000000  983931.000000  1.664089e+06  891586.000000   

                 X16            X17  
count   30000.000000   30000.000000  
mean    40311.904067   38873.869033  
std     60795.051774   59552.714581  
min    -81334.000000 -339603.000000  
25%      1769.500000    1258.500000  
50%     18107.000000   17081.000000  
75%     50177.000000   49198.250000  
max    927171.000000  961664.000000

As we can see in our table, the values mean values slightly changed compared to our original dataframe. However, the changes are really minimum. For example, in X12 the mean value decreased by approx  .019%. This change is really insignificant, therefore we conclude that using the median is a good way to replace the null values.

### COLUMN X18-X23 (AMOUNT OF PREVIOUS PAYMENT FROM APRIL TO SEPTEMBER, 2005)<br>
Amount of previous payment means to pay off the due balance of a previous month. This amount **should be zero or positive values.**<br>
So, first we are going to verify if there are negative values.


In [ ]:
dataframe[['X18','X19','X20','X21','X22','X23']].describe()

X18           X19            X20            X21  \
count   29992.000000  2.999100e+04   29992.000000   29989.000000   
mean     5662.945886  5.922489e+03    5225.623400    4827.252526   
std     16564.165089  2.304418e+04   17608.422625   15668.751975   
min         0.000000  0.000000e+00       0.000000       0.000000   
25%      1000.000000  8.355000e+02     390.000000     296.000000   
50%      2100.000000  2.009000e+03    1800.000000    1500.000000   
75%      5006.000000  5.000000e+03    4505.500000    4014.000000   
max    873552.000000  1.684259e+06  896040.000000  621000.000000   

                 X22            X23  
count   29989.000000   29995.000000  
mean     4800.297209    5216.259977  
std     15280.842069   17778.848359  
min         0.000000       0.000000  
25%       251.000000     118.000000  
50%      1500.000000    1500.000000  
75%      4033.000000    4000.000000  
max    426529.000000  528666.000000

We can see that there are not negative values, that is a good signal. Now we are going to see if we have NaN

In [ ]:
dataframe[['X18','X19','X20','X21','X22','X23']].isnull().any()

X18    True
X19    True
X20    True
X21    True
X22    True
X23    True
dtype: bool

We have NaN in all the columns.<br>
To solve this issue we will assume that the clients actually paid their exact amount stated in their bill statement from the previous month. In that regard, the strategic will be to copy the monthly balance of the bill statement of the previous month, as an example: <br>

**The balance of August will be the payment in September,**<br>
**The balance of July will be the payment in August, etc**<br>
This will just apply to **X18-X22**

In [ ]:
dataframe.X18.fillna(dataframe.X13, inplace=True)
dataframe.X19.fillna(dataframe.X14, inplace=True)
dataframe.X20.fillna(dataframe.X15, inplace=True)
dataframe.X21.fillna(dataframe.X16, inplace=True)
dataframe.X22.fillna(dataframe.X17, inplace=True)

In [ ]:
dataframe[['X18','X19','X20','X21','X22','X23']].isnull().any()

X18    False
X19    False
X20    False
X21    False
X22    False
X23     True
dtype: bool

As we can see above we dont have more NaN from X18 to X22. <br>
Now lets work on X23. As we can see below there are just 5 rows that has NaN values. We are going to delete the entire rows. 

In [ ]:
dataframe[dataframe['X23'].isnull()]

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
246      247  250000  2.0  2.0  1.0  75.0  0.0  0.0  0.0  0.0  ...   1010.0   
247      248  100000  2.0  2.0  2.0  27.0  0.0  0.0  0.0  0.0  ...  30337.0   
6232    6233   60000  2.0  2.0  2.0  29.0  2.0  2.0  2.0  0.0  ...  19052.0   
29824  29825   40000  1.0  1.0  1.0  47.0  2.0  2.0  2.0  2.0  ...  19052.0   
29825  29826   50000  1.0  2.0  1.0  41.0  0.0  0.0  0.0  0.0  ...  19052.0   

           X16      X17      X18      X19      X20      X21      X22  X23    Y  
246     5572.0    794.0   1631.0   1536.0   1010.0   5572.0    794.0  NaN  0.0  
247    30997.0  32904.0   1788.0   1799.0   1100.0   1150.0   2423.0  NaN  0.0  
6232   18107.0  17081.0  42117.0  20089.0  19052.0  18107.0  17081.0  NaN  NaN  
29824  18107.0  17081.0  21207.0  20089.0  19052.0  18107.0  17081.0  NaN  NaN  
29825  18107.0  17081.0  21207.0  20089.0  19052.0  18107.0  17081.0  NaN  NaN  

[5 rows x 25 columns]

In [ ]:
#Calculate how much % represent to delete 5 rows
100*(1 - dataframe.shape[0]/30000)
#dataframe9.info()

0.0

5 Rows it is just the 0.0166 % of the total database. You we are going to keep this strategic

In [ ]:
dataframe[['X23']].describe() # we run thsi to evaluate how much it affected our central tendencies after our deletion.

X23
count   29995.000000
mean     5216.259977
std     17778.848359
min         0.000000
25%       118.000000
50%      1500.000000
75%      4000.000000
max    528666.000000

In [ ]:
dataframe=dataframe.dropna()
dataframe[['X23']].describe()

X23
count   29995.000000
mean     5216.259977
std     17778.848359
min         0.000000
25%       118.000000
50%      1500.000000
75%      4000.000000
max    528666.000000

There was no change in the measurements of central tendency, it makes sense since NaN should not be consider when calculating this parameters. 

In [ ]:
dataframe.isnull().any()

ID     False
X1     False
X2     False
X3     False
X4     False
X5     False
X6     False
X7     False
X8     False
X9     False
X10    False
X11    False
X12    False
X13    False
X14    False
X15    False
X16    False
X17    False
X18    False
X19    False
X20    False
X21    False
X22    False
X23    False
Y      False
dtype: bool

**As we can see we dont have more NaN in our database**


##**Parte 3: Preparación de los datos**

###1. ¿Qué datos consideró más importantes? ¿Por qué?

According to the bibliography of this dataset, around 2009, banks in Taiwan faced an economic crisis that impacted on the bank’s confidence over their clients’ finances. This crisis was caused by the banks over-issuing cash and credit cards to unqualified applicants. This dataset was created with historical data from this event and its objective was to create a predictive model that would classify the future clients as credible or not credible clients. In other words, it was created from a risk management perspective.\
With this in mind, we can consider that some sets or columns in the database are more important than others because they help us predict with more accuracy whether a future client is reliable or not. Specifically, we believe that the variables of X1 (Amount of the given credit, X6-X11 (History of past payment), X12-X17 (Amount of bill statement), and X18-X23 (Amount of previous payment).
As for the rest of the data, we believe it is helpful to draw descriptive conclusions like a married man might be more likely to fail on payments that a single man. However, they are not as decisive to predict reliability. 

###2. ¿Se eliminaron o reemplazaron datos nulos? ¿Qué se hizo y por qué?

First, it is important to mention that in this exercise and as a first approach, we decided to treat every column as an independent variable. Therefore, we analyzed individually each column. 

During this analysis, there was only one column in which its null values were deleted. For the rest of the columns, their null values were replaced. We believe that both ways were a good approach because the first insight that we had from the data was that the null values missing per column was minimal. For example, Column X16 has the most null values but they were only 17 out of 30,000. This would result in deleting 0.05% of the rows out of the total sample. However, if we deleted the null values of all columns, this would result in deleting around 50 rows out of 30,0000. Although it still seems that 50 out of 30,000 is still a minimal number, if we consider that the dataset’s objective is to create a predictive model then, 30,000 rows might not be big enough. Therefore, we decided to keep deletions to the minimum and, instead, decided to replace most null values. 

When replacing values, we had different approaches. There were some columns that their null values were integrated to specific groups (in the educational levels, NaN were integrated into the 4=others category), and other columns where the null values were replaced by the median. It is important to mention that replacing null values didn’t seem to have a big impact on our data distribution based on measurements of central tendency. You can see that when we replaced the null values in Columns X12-X17, the mean values slightly changed compared to our original dataset. However, the changes are minimum. For example, in X12 the mean value decreased by approx. .019%. 

For columns X18-X23, we decided to assume that that the clients actually paid the exact amount stated in their bill statement from the previous month. In that regard, the strategic was to copy the monthly balance on the amount of payment for previous month, as an example: The balance of August will be the payment in September, the balance of July will be the payment in August, and so on. However, this was not applicable for column X23, since the payment made in April corresponded to the bill statement from March and we did not have it. Since we couldn’t follow the same logic, we decided to delete the rows with null values. Only 5 rows were deleted which represented 0.0166 % of the total database.

Finally, when we deleted these 5 rows, we automatically also deleted the three rows that had NaN values in the Y column. This was pure luck, but helped us. 

###3. ¿Es necesario ordenar los datos para el análisis? Sí / No / ¿Por qué?

Data sorting is any process related to arranging the data into a meaningful way with the objective of making it easier to understand, analyze or visualize. In this case, we don’t think that the data must be in a specific order because, since we are talking about independent clients, then there is no sequential order that the data must follow. 
However, as you can see in our code, when analyzing the data, we did sort our columns into categories. There were mainly two categories that we used:

*	**Categorical Columns:** these were variables that had  a range of specific numerical values that should be used and specific categorical values for each numerical value. For example, in gender 1 = male; 2 = female.
* **Numerical Columns:** these were variables which values had no categorical interpretation. They were pure numerical values. 

We decided to divide our dataset this way to have a better visualization of the descriptive table. This way we were able to easily spot format errors or erroneous data in the categorical data that will be described in the following question.


###4. ¿Existen problemas de formato que deban solucionar antes del proceso de modelado? Sí / No / Por qué.

There were some format issues that were solved during the first data analysis. We can make this statement mainly because we noticed that on the categorical variables there were undefined values in the data set that needed to match the defined format. As described in question 3, all these columns had a specific defined scale that should’ve been used in the dataset. However, we found that all of them had data format errors or erroneous data.

As you noticed in our code, we classified the following variables as categorical and solved their problems in several ways:

*	**X2 (GENDER):** this had no format error. Only a null value that was modified based on the median value of the column.

*	**X3 (EDUCATION):** For this column we should only have four values (1 = graduate school; 2 = university; 3 = high school; 4 = others). However, in our analysis we realize that we had three values out of the scale (5,6,0) and NaNs. Since, in this case we know that 4 = others, then we decided to store the undefined values (5, 6, 0) as a 4 to match the column format. 

*	**X4 (MARITAL STATUS):** This was a similar situation. For this column we should only have three values (1 = married; 2 = single; 3 = others). However, in our analysis we realize that we had one value out of the scale (0) and NaNs. Since, in this case we know that 3 = others, then we decided to store the undefined values (0 and NaN) as a 3 to match the column format.

*	**X6-X11 (PAYMENT STATUS):** All these columns should have the following values -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above. During our analysis, we noticed that we had two values out of scale (0, -2) and NaNs. As a solution, we thought convenient to group all unwanted data (0, -2 and NaN) inside the category 'pay duly', since 0 and -2 probably means that they did not have a delay on the payment. Also, we could define 0 as "pay duly" instead of -1 that way we have a sequence of numbers.


###5. ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?

As explained throughout these questions, we used several methods depending on the columns:

* **X2 (GENDER):** we decided to replace the null values with the median (2.0) of the column.
*	**X3 (EDUCATION):** We decided to integrate the undefined and null values into the “Others” category. 
*	**X4 (MARITAL STATUS):** We decided to integrate the undefined and null values into the “Others” category.
*	**X6-X11 (PAYMENT STATUS):** We decided to redefine the attributes of the column (“pay duly” was defined as 0 instead of -1). Also, all undefined and null values were stored in 0.
*	**X5 (AGE):** We decided to replace the missing values with the median (34.0) of the column.
* **X12-X17 (AMOUNT OF BILL STATEMENT):** We decided to replace the missing values with the median of each column. 
* **X18-X22 (AMOUNT OF PREVIOUS PAYMENT):** We decided to replace the missing values with the monthly bill statement of the previous month. 
*	**X23 (AMOUNT OF PREVIOUS PAYMENT):** We decided to delete the rows that had null values in this column. 







##**Referencias**

Kumar, A. (2021). Python - Replace Missing Values with Mean, Median & Mode - Data Analytics. Data Analytics. Retrieved 4 October 2022, from https://vitalflux.com/pandas-impute-missing-values-mean-median-mode/#:~:text=df.mean())-,Impute%20%2F%20Replace%20Missing%20Values%20with%20Median,for%20replacing%20the%20missing%20values.

MIT. (2022). I have a negative balance on my bill. What does this mean? | MIT Student Financial Services. MIT Student Financial Services. Retrieved 4 October 2022, from https://sfs.mit.edu/help/faq/i-have-a-negative-balance-on-my-bill-what-does-this-mean/.

ORACLE. (2005). Sorting data. Docs.oracle.com. Retrieved 4 October 2022, from https://docs.oracle.com/cd/B14099_19/bi.1012/b13915/sorting.htm.

Yeh, I. C., & Lien, C. H. (2009). The comparisons of data mining techniques for the predictive accuracy of probability of default of credit card clients. Expert systems with applications, 36(2), 2473-2480.


